### PR TITLE
feat: Member list - with controller

### LINKED ISSUE #13: Visitors should be able to see members of the site.
A new "Members" page should be present that displays a paginated list of all users within the system. 

They should be presented with a search for this page only that searches and filters the list of users by the following criteria: 
- username
- role (admin, moderator especially)
- new users (with 1 or less posts)
- country

You should be able to sort the list of users by : 
- number of posts
- username, asc
- most recent posts, desc

### PR DESCRIPTION
**Pull request by michalsn**
An alternative version to #26 - this time using a more standard approach: the controller.

Closes #13

### CODE REVIEW

File: app/Concerns/ImpactsForumCounts.php [hunk 1]
```diff
@@ -18,10 +19,20 @@ protected function incrementThreadCount(array $data)
...
+        $userModel = model(UserModel::class);
+
+        // Increment Forum thread count
+        $forum = $forumModel->find($thread->forum_id);
         $forum->thread_count++;
 
-        model(ForumModel::class)->save($forum);
+        $forumModel->save($forum);
+
+        // Decrement User thread count
```
Comment (lonnieezell) [inline app/Concerns/ImpactsForumCounts.php]:
```suggestion
        // Increment User thread count
```

File: app/Models/UserModel.php [hunk 2]
```diff
@@ -2,10 +2,85 @@
...
+        $selects = [
+            'users.*',
+            '(thread_count + post_count) AS count',
+        ];
+
+        $results = $this
+            ->withIdentities()
+            ->select(implode(', ', $selects))
+            ->join('auth_groups_users', 'auth_groups_users.user_id = users.id', 'left')
+            ->when(isset($search['username']) && $search['username'] !== '', static function ($query) use ($search) {
```
Comment (lonnieezell) [inline app/Models/UserModel.php]: Not saying you need to change it, but newer versions of PHP 8 allow new function types that are scoped to the outer function, like: 

```suggestion
            ->when(isset($search['username']) && $search['username'] !== '', fn ($query) => $query->like('users.username', $search['username'], 'both');
```

Or something like that.

**Review by lonnieezell — APPROVED** [review]
A couple of notes in there, but I like going this way instead of the cell better. I think it's more straight-forward and a better practice for people looking at our code. 

Thanks!

File: app/Models/UserModel.php (hunk 2 above)
Comment (michalsn) [reply app/Models/UserModel.php]: Will do. I've recently spent some time with older PHP 7 projects, so the new syntax doesn't come naturally.

File: app/Views/members/list.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                                </div>
+                                <div class="px-4">
+                                    <div class="font-bold cursor-pointer" hx-get="<?= esc($member->link(), 'attr') ?>"><?= esc($member->username); ?></div>
+                                    <div class="text-sm opacity-50"><?= is_array($member->roles) ? implode(', ', $member->roles) : $member->roles; ?></div>
+                                </div>
+                            </div>
+                        </td>
+                        <td><?= $member->count; ?></td>
+                        <td><?= esc($member->country); ?></td>
+                        <td><?= $member->last_active->humanize() ?? 'Never'; ?></td>
```
Comment (datamweb) [inline app/Views/members/list.php]:
```suggestion
                        <td><?=($member->last_active !== null) ? $member->last_active->humanize() : 'Never';?></td>
```

File: app/Controllers/Members/MemberController.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            'sortDirection'   => ['in_list[asc,desc]'],
+        ];
+
+        if (! $this->validateData($table, $rules)) {
+            throw new InvalidArgumentException(implode(PHP_EOL, $this->validator->getErrors()));
+        }
+
+        helper('form');
+
+        $model = model(UserModel::class);
```
Comment (datamweb) [inline app/Controllers/Members/MemberController.php]:
```suggestion
        $userModel= model(UserModel::class);
```

File: app/Controllers/Members/MemberController.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        $data = [
+            'members' => $model->searchMembers(
```
Comment (datamweb) [inline app/Controllers/Members/MemberController.php]:
```suggestion
            'members' => $userModel->searchMembers(
```

File: app/Controllers/Members/MemberController.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                $table['search'], $table['page'], $table['perPage'], $table['sortColumn'], $table['sortDirection']
+            )
+        ];
+
+        $table['dropdowns'] = [
+            'role' => array_combine($roleKeys, array_merge(['All'], array_column(setting('AuthGroups.groups'), 'title'))),
+            'type' => ['all' => 'All users', 'new' => 'New users'],
+        ];
+
+        $table['pager']  = $model->pager;
```
Comment (datamweb) [inline app/Controllers/Members/MemberController.php]:
```suggestion
        $table['pager']  = $userModel->pager;
```

**Review by datamweb — APPROVED** [review]
@michalsn thank you!
LGTM!

**Comment by michalsn** [comment]
Thanks, everyone.

**Pull request merged**